File: 290. Word Pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        str = str.split(' ')
        if len(str) != len(pattern):
            return False
        words_to_patterns = {}
        patterns_to_words = {}
        for i,j in zip(str, pattern):
            if i not in words_to_patterns and j not in patterns_to_words:
                words_to_patterns[i] = j
                patterns_to_words[j] = i
            else:
                if i in words_to_patterns and j in patterns_to_words and words_to_patterns[i] == j and patterns_to_words[j] == i:
                    continue
                else:
                    return False
        return True

File: 290. Word Pattern.cpp
#include <sstream>
#include <unordered_map>
class Solution {
public:
    vector<string> split(string str, char delimiter)
    {
        vector<string> internal;
        stringstream ss(str);
        string tok;
        while(getline(ss, tok, delimiter))
            internal.push_back(tok);
        return internal;
    }
    bool wordPattern(string pattern, string str) {
        vector<string> str_split = split(str, ' ');
        if (pattern.length() != str_split.size())
            return false;
        unordered_map <string, char> words_to_patterns;
        unordered_map <char, string> patterns_to_words;
        for (int i = 0; i < pattern.size(); i++)
        {
            if (words_to_patterns.count(str_split[i]) == 0 && patterns_to_words.count(pattern[i]) == 0)
            {
                words_to_patterns[str_split[i]] = pattern[i];
                patterns_to_words[pattern[i]] = str_split[i];
            }
            else
            {
                if (words_to_patterns.count(str_split[i]) > 0 && patterns_to_words.count(pattern[i]) > 0 && words_to_patterns[str_split[i]] == pattern[i] && patterns_to_words[pattern[i]] == str_split[i])
                    continue;
                else
                    return false;
            }
        }
        return true;
    }
};
